### ISSUE TITLE
SQL logic error when starting a ValueObservation immediately after opening a DatabasePool

### ISSUE THREAD

**Issue by kustra** [groue/GRDB.swift#1383]
### TL;DR

GRDB v6.14.0 fixes this issue. If you can't upgrade, and your app doesn't perform SQLite checkpoints, here's a workaround - to be run right after creating the `DatabasePool`:

        let dbPool: DatabasePool = ... // your init code here
        
        // https://github.com/groue/GRDB.swift/issues/1383
        // FIXME: remove this after upgrading to GRDB v6.14.0
        try dbPool.writeWithoutTransaction { db in
            try db.inSavepoint {
                try db.execute(sql: """
                CREATE TABLE grdb_issue_1383 (id INTEGER PRIMARY KEY);
                DROP TABLE grdb_issue_1383;
                """)
                return .commit
            }
        }

### What did you do?

I created a new `DatabasePool` with an existing .sqlite database file that did not have -wal or -shm files next to it. I noticed that by the time the `DatabasePool` was created, there was already a zero-length -wal file and an -shm file next to the .sqlite file.

I debugged into it, and executing this line creates the -wal and -shm files: https://github.com/groue/GRDB.swift/blob/v6.13.0/GRDB/Core/Database.swift#L420
Notably, this `if` condition evaluated to `false`, because a zero-length -wal file existed at this point: https://github.com/groue/GRDB.swift/blob/v6.13.0/GRDB/Core/DatabasePool.swift#L87

Then, I tried to start a ValueObservation with the `DatabasePool`, which failed on this line: https://github.com/groue/GRDB.swift/blob/v6.13.0/GRDB/Core/WALSnapshot.swift#L45

I suspect it fails because of [this requirement](https://www.sqlite.org/c3ref/snapshot_get.html):
> One or more transactions must have been written to the current wal file since it was created on disk (by any connection). This means that a snapshot cannot be taken on a wal mode database with no wal file immediately after it is first opened. At least one transaction must be written to it first.

If I place a breakpoint at the end of the `DatabasePool` init, and execute this command from the debugger, the `ValueObservation` seems to work as expected: https://github.com/groue/GRDB.swift/blob/v6.13.0/GRDB/Core/DatabasePool.swift#L93

I'm not sure if this is important, but the .sqlite file was first created using GRDB 5.12.0, then I tried opening it with GRDB 6.13.0.

### What did you expect to happen?

I expect a `ValueObservation` to work even if I start it right after opening a `DatabasePool`.

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:** 6.13.0
**Installation method:** CocoaPods
**Xcode version:** 14.2
**Swift version:** 5.7.2
**Platform(s) running GRDB:** iOS
**macOS version running Xcode:** 12.6.5

**Comment by groue** [comment]
> Yes, this is expected to work.

@kustra, I'm afraid this can't be done reliably by GRDB 😬 

The root problem is that creating a wal snapshot from a truncated wal file is not supported by SQLite. Wal snapshots are used by `ValueObservation` but also [`DatabaseSnapshotPool`](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/databasesnapshotpool).

Ideally, we'd recreate a wal file, whenever it is empty, before creating a wal snapshot.

Unfortunately, the concurrency model of `DatabasePool` can't correctly deal with this. A pool serializes writes but allows concurrent reads and writes. Wal snapshots belong to the realm of reads, while truncating checkpoints belong to the realm of writes. Users can perform both whenever they want. In the concurrent world of a database pool, it is impossible to guarantee that no truncating checkpoint is performed before a wal snapshot is created: they are concurrent and there will always be a chance that something wrong happens, i.e. that someone opens a similar issue (but essentially impossible to fix).

I don't want to ship code that makes such an impossible promise, so the only solution will be to make applications responsible for the scheduling of checkpoints and observations. Applications will have to make sure the wal file exists before they start an observation. I'll come up with an api. Suggestions are welcome, of course :-)

**Comment by kustra** [comment]
I think the most common use case is that the app initiates no checkpoints, and the developer doesn't know about this requirement for a ValueObservation. Perhaps updating the documentation of `ValueObservation` is enough, along with making this common use case work out of the box. Assuming that SQLite doesn't perform truncating checkpoints on its own.

**Comment by groue** [comment]
Yes, I agree. I was contemplating degrading the ValueObservation if it fails creating a snapshot, and make it use the code path activated on older versions of the operating system, where WAL snapshots are not available.

This "degraded" mode has a drawback: the observation always start by notifying _two_ values, even if the database was not changed. For more information, see #1248, and [this comment](https://github.com/groue/GRDB.swift/pull/1364#issuecomment-1512637667) that describes the evolution of ValueObservation on DatabasePool.

**Comment by groue** [comment]
By the way, does your app perform truncating checkpoints, or did you just meet a truncated WAL file by chance?

**Comment by kustra** [comment]
> By the way, does your app perform truncating checkpoints, or did you just meet a truncated WAL file by chance?

I also fall into what I guess is the most common use case: I didn't know about SQLite checkpoints at all until this afternoon. I learned a lot more about them today than I intended to. :) My app doesn't perform checkpoints. I just debugged why the app started producing errors around ValueObservations on launch, after upgrading GRDB.

Also, I don't understand why it's not happening for a lot more people. It happens every time I open an .sqlite file and start a ValueObservation before writing anything to the database.

**Comment by kustra** [comment]
> Well, thank you very much for your investigation, because your initial report has fully identified the problem 👍 That's very much appreciated 🙂

You're welcome.


> > Also, I don't understand why it's not happening for a lot more people. It happens every time I open an .sqlite file and start a ValueObservation before writing anything to the database.
> 
> I don't understand either. This error never happened to me, and nobody has reported it yet.

But [didn't you just create](https://github.com/groue/GRDB.swift/issues/1383#issuecomment-1572224232) a [failing unit test](https://github.com/groue/GRDB.swift/commit/d66d2118b5fa4c17b185301db68eb7ae5727feb9) that reproduces this issue only using GRDB?
Btw I think this unit test could be simplified further. There's no need to open the .sqlite file twice, or execute a checkpoint. Let me know if you have trouble reproducing the bug like that and I'll try with my .sqlite file.

I could also reproduce the issue from a "clean" state where no -wal file is present next to the .sqlite file, only using GRDB. The QA team member that initially found this bug definitely didn't touch the .sqlite file with anything but our app.

**Comment by groue** [comment]
> But [didn't you just create](https://github.com/groue/GRDB.swift/issues/1383#issuecomment-1572224232) a [failing unit test](https://github.com/groue/GRDB.swift/commit/d66d2118b5fa4c17b185301db68eb7ae5727feb9) that reproduces this issue only using GRDB?

Yes! This issue is the first report of this error, so it's the first opportunity to write a regression test :-)

> Btw I think this unit test could be simplified further. There's no need to open the .sqlite file twice, or execute a checkpoint.

The checkpoint is necessary in order to have a zero-sized wal file (and trigger the [fourth condition](https://www.sqlite.org/c3ref/snapshot_get.html) that prevents the creation of a wal snapshot). The second connection may not be necessary but 1. for some unclear reason I can't make the test fail without it, and 2. this second connection make the test closer to your own setup (as far as I understand).

> I could also reproduce the issue from a "clean" state where no -wal file is present next to the .sqlite file, only using GRDB. 

🤔 That's probably another issue, because the -wal file is created by DatabasePool initializer if it is missing... I'm interested in a reproducing sample code :-)

**Comment by kustra** [comment]
> 🤔 That's probably another issue, because the -wal file is created by DatabasePool initializer if it is missing... I'm interested in a reproducing sample code :-)

[Do you mean that this block creates the -wal file on DatabasePool init?](https://github.com/groue/GRDB.swift/blob/v6.13.0/GRDB/Core/DatabasePool.swift#L87)
If you create a DatabasePool with an .sqlite file that has no -wal file next to it, this block won't run, because [this line of code](https://github.com/groue/GRDB.swift/blob/v6.13.0/GRDB/Core/Database.swift#L420), called by DatabasePool.init, runs earlier, and automatically creates a zero-length -wal file.
You can reproduce it by removing the -wal file after a truncating checkpoint (this is done by default by the `sqlite3` command line tool), then opening it with a DatabasePool.

**Comment by groue** [comment]
> [Do you mean that this block creates the -wal file on DatabasePool init?](https://github.com/groue/GRDB.swift/blob/v6.13.0/GRDB/Core/DatabasePool.swift#L87)

Yes, it does.

> If you create a DatabasePool with an .sqlite file that has no -wal file next to it, this block won't run, because [this line of code](https://github.com/groue/GRDB.swift/blob/v6.13.0/GRDB/Core/Database.swift#L420), called by DatabasePool.init, runs earlier, and automatically creates a zero-length -wal file.

Interesting. Maybe the first line is no longer needed, then. I'll check that.

Now, the net effect is the same: the wal file *exists* (even if it is empty), so this is not a distinct issue.

**Comment by groue** [comment]
> It is not necessary. I could reproduce the error with this test:

I stand corrected.

So this code creates an empty wal file?! 😮

But you certainly shake my intuitions. And now I really understand your question about the frequency of this error. 🤯

**Comment by kustra** [comment]
> Now, the net effect is the same: the wal file *exists* (even if it is empty) after the DatabasePool has been instantiated, so this is not a distinct issue.

That's not true. If the -wal file is created by the validateFormat() call, ValueObservations fail. If it's created by the CREATE TABLE grdb_issue_102 command in DatabasePool.init, ValueObservations work. I guess it's because the former is just a read operation, and the latter is writing to the database.

**Comment by groue** [comment]
It looks like something on your setup has `validateFormat()` create the wal file. This is not the case in my tests, where the `CREATE TABLE grdb_issue_102` is necessary for the wal file to be created, despite the call to `validateFormat()`. That's weird.

Would you please tell me if your various setup still fail with the latest [dev/issue-1383](https://github.com/groue/GRDB.swift/tree/dev/issue-1383) branch?

**Comment by kustra** [comment]
> It looks like something on your setup has `validateFormat()` create the wal file. This is not the case in my tests, where the `CREATE TABLE grdb_issue_102` is necessary for the wal file to be created, despite the call to `validateFormat()`. That's weird.

I can confirm. When creating an .sqlite file from the unit test, calling `validateFormat()` on it doesn't create the -wal file.

[I have attached a sample .sqlite file.](https://github.com/groue/GRDB.swift/files/11630296/Issue1383.sqlite.zip) I created it by taking an .sqlite file from my app, [wiping it](https://stackoverflow.com/a/65743498), then running `CREATE TABLE t(a)` in the `sqlite3` command line tool. There's no attending -wal file. This test fails with it, because `validateFormat()` creates the -wal file in a way that doesn't work with ValueObservations on v6.13.0:

    func testIssue1383_createWal() throws {
        let url = testBundle.url(forResource: "Issue1383", withExtension: "sqlite")!
        let dbPool = try DatabasePool(path: url.path, configuration: dbConfiguration)
        let observation = ValueObservation.tracking(Table("t").fetchCount)
        _ = observation.start(
            in: dbPool, scheduling: .immediate,
            onError: { error in
                XCTFail("Unexpected error \(error)")
            },
            onChange: { _ in
            })
    }


> Would you please tell me if your various setup still fail with the latest [dev/issue-1383](https://github.com/groue/GRDB.swift/tree/dev/issue-1383) branch?

I'll get back to you on that tomorrow. I checked the ValueObservationTests, though, and those succeed on this branch on my machine, including the one above - though it probably passes by falling back to the alternative method that you've mentioned.

I feel like [this condition](https://github.com/groue/GRDB.swift/blob/v6.13.0/GRDB/Core/DatabasePool.swift#L87) could be changed to also execute if the -wal file exists, but is zero-length. The code inside doesn't cause harm, even if in some cases, it executes unnecessarily.

**Comment by groue** [comment]
> I have attached a sample .sqlite file.

Thank you. I added a test for it. Indeed `validateFormat()` creates an empty wal file in this case. The test pass with the fix.

> I feel like [this condition](https://github.com/groue/GRDB.swift/blob/v6.13.0/GRDB/Core/DatabasePool.swift#L87) could be changed to also execute if the -wal file exists, but is zero-length. The code inside doesn't cause harm, even if in some cases it executes unnecessarily.

This change does not aim precisely at our issue. It would not fix observations started from a truncated wal file in general. This can happen when a truncating checkpoint is performed after the DatabasePool instance has been created.

But is has some merit: ValueObservation would not need to run in degraded mode (avoid wal snapshots) for apps that do not perform truncating checkpoints, but are unlucky enough to boot from a truncated wal file.

So yes, this is a good idea 👍

**Comment by groue** [comment]
> So yes, this is a good idea

Done in the [dev/issue-1383](https://github.com/groue/GRDB.swift/tree/dev/issue-1383) branch.

**Comment by kustra** [comment]
I just checked [dev/issue-1383](https://github.com/groue/GRDB.swift/tree/dev/issue-1383) in my app. It works.
Awesome. Thank you for the quick response, and in general, for maintaining GRDB! :thumbsup: :clap: :rocket:

**Comment by groue** [comment]
You have been a super user, too, very precise and challenging! It was a pleasure to deal with this issue with you, and quickly find a fix that will help all other users 🤝 I hope you'll find other issues to chew on 😉

#1384 should be merged and shipped soon.
